### PR TITLE
#9749: move repeat

### PR DESCRIPTION
**Pull request by ntarafdar**
### Ticket
[Link to Github Issue](https://github.com/tenstorrent/tt-metal/issues/9749)

### Problem description
Repeat is in tt lib moving to ttnn

### What's changed
Migrate all uses and implementation from tt lib to ttnn

### Checklist
- [ ] Post commit CI passes
- [ ] Blackhole Post commit (if applicable)
- [ ] Model regression CI testing passes (if applicable)
- [ ] New/Existing tests provide coverage for changes

### CODE REVIEW

File: ttnn/cpp/ttnn/operations/data_movement/repeat/repeat_pybind.hpp [hunk 1]
```diff
@@ -0,0 +1,62 @@
+// SPDX-FileCopyrightText: © 2023 Tenstorrent Inc.
+//
+// SPDX-License-Identifier: Apache-2.0
+
+#pragma once
```
Comment (ayerofieiev-tt) [inline ttnn/cpp/ttnn/operations/data_movement/repeat/repeat_pybind.hpp:5]: please split similarly to ccp/all_gather
* move impl to cpp
* reduce headers to pybind_fwd.hpp

**Review by ayerofieiev-tt — APPROVED** [review]

File: models/demos/falcon7b_common/tt/falcon_model.py [hunk 2]
```diff
@@ -159,7 +159,7 @@ def model_preprocessing(self, llm_mode, input_ids, kv_cache_len, num_input_token
                 )
                 # Repeat attn masks for all heads
                 for i in range(self.num_devices):
-                    tt_attention_mask[i] = ttnn.experimental.tensor.repeat(
+                    tt_attention_mask[i] = ttnn.repeat(
                         tt_attention_mask[i],
                         [1, self.config.num_attention_heads, 1, 1],
                         output_mem_config=self.model_config["ATTN_MASK_MEMCFG"],
```
Comment (skhorasganiTT) [inline models/demos/falcon7b_common/tt/falcon_model.py:165]: should this be memory_config?

Comment (tarafdarTT) [reply models/demos/falcon7b_common/tt/falcon_model.py:165]: oops yes , fixing now

**Review by skhorasganiTT — APPROVED** [review]

File: ttnn/cpp/ttnn/operations/data_movement/repeat/repeat.hpp [hunk 3]
```diff
@@ -0,0 +1,34 @@
...
+        const Shape & repeat_dims,
+        const std::optional<MemoryConfig>& memory_config);
+
+    static ttnn::Tensor operator()(const ttnn::Tensor& input_tensor, const Shape & repeat_dims);
+};
+
+
+}  // namespace operations::data_movement
+
+constexpr auto repeat = ttnn::register_operation<"ttnn::repeat", ttnn::operations::data_movement::RepeatOperation>();
```
Comment (sjameelTT) [inline ttnn/cpp/ttnn/operations/data_movement/repeat/repeat.hpp:32]: Does this support async? ttnn::register_operation_with_auto_launch_op then

**Review by sjameelTT — APPROVED** [review]

File: ttnn/cpp/ttnn/operations/data_movement/repeat/device/repeat_op.cpp [hunk 4]
```diff
@@ -0,0 +1,54 @@
...
+
+
+void RepeatDeviceOperation::validate(const std::vector<Tensor> &input_tensors) const {
+    const auto &input_tensor = input_tensors[0];
+    tt::tt_metal::Shape input_shape = input_tensor.get_legacy_shape();
+    TT_FATAL(this->repeat_dim < input_shape.rank(), "Repeat dim specified is larger than input tensor rank.");
+    if (input_tensor.get_layout() == Layout::ROW_MAJOR && this->repeat_dim == input_shape.rank() - 1) {
+        TT_FATAL(
+            (input_shape[this->repeat_dim] * input_tensor.element_size()) % input_tensor.buffer()->alignment() == 0,
+            "Current repeat implementation requires aligned last dim when repeating on last dim");
```
Comment (dmakoviichuk-tt) [inline ttnn/cpp/ttnn/operations/data_movement/repeat/device/repeat_op.cpp:22]: Could you provide better error message.

**Review by dmakoviichuk-tt — APPROVED** [review]
Please review your TT_FATAL error messages you can make them more actionable. 
Overall LGTM

**Review by cglagovichTT — APPROVED** [review]

**Review by eyonland — APPROVED** [review]

**Review by arakhmati — APPROVED** [review]

**Review by cfjchu — APPROVED** [review]

**Review by TT-BrianLiu — APPROVED** [review]

**Pull request merged**